### PR TITLE
Add structured trace logs

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #120731
* #120658
* __->__ #120289
* #120712

Overall design: https://docs.google.com/document/d/1CX_hJ0PNy9f3R1y8TJrfkSeLkvGjjjLU84BSXgS2AZ8/edit

How to read the diff:
* Most files are me augmenting pre-existing logging with structured variants. For the most part it's simple (esp FX graphs, which have a canonical string representation); it gets more complicated when I decided to JSON-ify some data structure instead of keeping the ad hoc printing (notably, guards and dynamo output graph sizes)
* torch/_functorch/_aot_autograd/collect_metadata_analysis.py is some unrelated fixes I noticed while auditing artifact logs
* torch/_logging/_internal.py has the actual trace log implementation. The trace logger is implement as a logger named torch.__trace which is disconnected from the logging hierarchy. It gets its own handler and formatter (TorchLogsFormatter with _is_trace True). `trace_structured` is the main way to emit a trace log. Unusually, there's a separate "metadata" and "payload" field. The metadata field should not be too long (as it is serialized as a single line) and is always JSON (we put contextual things like compile id in it); the payload field can be long and is emitted after the metadata log line and can span multiple lines.
* torch/_logging/structured.py contains some helpers for converting Python data structures into JSON form. Notably, we have a string interning implementation here, which helps reduce the cost of serializing filenames into the log.
* test/dynamo/test_structured_trace.py the tests are cribbed from test_logging.py, but all rewritten to use expect tests on munged versions of what we'd actually output. Payloads are never tested, since they tend not be very stable.

https://github.com/ezyang/tlparse is a POC Rust program that can interpret these logs.

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames

### CODE REVIEW

File: torch/_dynamo/output_graph.py [hunk 1]
```diff
@@ -999,30 +999,38 @@ def cleanup_graph(self):
                     self.graph.erase_node(node1)
                     self.graph.erase_node(node2)
 
-    def get_graph_sizes_log_str(self, name):
-        graph_sizes_str = "TRACED GRAPH TENSOR SIZES\n"
-        graph_sizes_str += f"===== {name} =====\n"
+    def get_graph_sizes(self, *, structured, name=None):
```
Comment (Skylion007) [inline torch/_dynamo/output_graph.py:1002]: Maybe mypy is confused about the typing of the structured param? Is it a boolean?
```suggestion
    def get_graph_sizes(self, *, structured : bool, name : Optional[str] =None):
```

File: torch/_logging/structured.py [hunk 2]
```diff
@@ -0,0 +1,18 @@
+"""
+Utilities for converting data types into structured JSON for dumping.
+"""
+
+import traceback
+from typing import Sequence
+
+def from_traceback(tb: Sequence[traceback.FrameSummary]) -> object:
```
Comment (Skylion007) [inline torch/_logging/structured.py:8]:
```suggestion
def from_traceback(tb: Sequence[traceback.FrameSummary]) -> List[Dict[str, object]]:
```

**Review by Skylion007 — APPROVED** [review]

File: torch/_logging/structured.py (hunk 2 above)
Comment (ezyang) [reply torch/_logging/structured.py:8]: due to mutability of list/dict, this doesn't type check!

**Comment by ezyang** [comment]
The /logs integration bit needs a killswitch

**Comment by ezyang** [comment]
I've dumped the fbcode special code, and will be landing that separately.

File: torch/_logging/_internal.py [hunk 3]
```diff
@@ -865,6 +900,63 @@ def _init_logs(log_file_name=None):
...
+                logging.StreamHandler.close(self)
+        finally:
+            self.release()
+
+    def emit(self, record):
+        if self.stream is None:
+            open_func = self._builtin_open
+            self.stream = open_func(self.filename_cb(), "w")
+        if self.stream:
+            super().emit(record)
```
Comment (ezyang) [inline torch/_logging/_internal.py:958]: this is dead code, but won't be dead later

**Pull request closed without merging**